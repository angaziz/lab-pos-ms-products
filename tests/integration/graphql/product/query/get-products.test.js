const { gql } = require('apollo-server-express');
const mockingoose = require('mockingoose').default;
const {
  mongo: {
    MongoError,
  },
} = require('mongoose');
const { testClient } = require('../../../../helpers');
const { Product } = require('../../../../../src/models');
const productListFixture = require('../../../../fixtures/models/products/product-list.json');

describe('[INTEGRATION] [GRAPHQL] [QUERY] - getProducts', () => {
  const { query } = testClient();
  const GET_PRODUCTS = gql`
    query ($offset: Int!, $limit: Int!) {
      getProducts(offset: $offset, limit: $limit) {
        id
        code
        name
        buyPrice
        sellPrice
        discountPercent
        quantity
        createdAt
        updatedAt
      }
    }
  `;

  test('Should return expected data when request is success', async () => {
    mockingoose(Product).toReturn(productListFixture, 'find');

    const res = await query({
      query: GET_PRODUCTS,
      variables: { offset: 0, limit: 5 },
    });

    expect(res.errors).toBeFalsy();
    expect(res.data.getProducts.length).toBe(productListFixture.length);
  });

  test('Should return expected error when error is occured', async () => {
    mockingoose(Product).toReturn(new MongoError('some error'), 'find');

    const res = await query({
      query: GET_PRODUCTS,
      variables: { offset: 0, limit: 5 },
    });

    expect(res.errors[0].constructor.name).toBe('GraphQLError');
  });
});
