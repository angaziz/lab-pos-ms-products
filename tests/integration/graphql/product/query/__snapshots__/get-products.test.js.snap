// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[INTEGRATION] [GRAPHQL] [QUERY] - getProducts Should return expected data when requested 1`] = `
Object {
  "data": Object {
    "getProducts": Array [
      Object {
        "buyPrice": 20,
        "code": "BOOK01",
        "createdAt": "1564232856176",
        "discountPercent": 0,
        "id": "5d3c4c980c79592fead7aaaa",
        "name": "Agile Principles - Robert C. Martin",
        "quantity": 10,
        "sellPrice": 30,
        "updatedAt": "1564232856176",
      },
      Object {
        "buyPrice": 20,
        "code": "BOOK02",
        "createdAt": "1564232856176",
        "discountPercent": 0,
        "id": "5d3c4c980c79592fead7bbbb",
        "name": "Refactoring - Martin Fowler",
        "quantity": 10,
        "sellPrice": 30,
        "updatedAt": "1564232856176",
      },
      Object {
        "buyPrice": 20,
        "code": "BOOK03",
        "createdAt": "1564232856176",
        "discountPercent": 0,
        "id": "5d3c4c980c79592fead7cccc",
        "name": "Domain Specific Language - Martin Fowler",
        "quantity": 10,
        "sellPrice": 30,
        "updatedAt": "1564232856176",
      },
      Object {
        "buyPrice": 20,
        "code": "BOOK04",
        "createdAt": "1564232856176",
        "discountPercent": 0,
        "id": "5d3c4c980c79592fead7dddd",
        "name": "Clean Architecture - Robert C. Martin",
        "quantity": 10,
        "sellPrice": 30,
        "updatedAt": "1564232856176",
      },
      Object {
        "buyPrice": 20,
        "code": "BOOK05",
        "createdAt": "1564232856176",
        "discountPercent": 0,
        "id": "5d3c4c980c79592fead7eeee",
        "name": "The Clean Coder - Robert C. Martin",
        "quantity": 10,
        "sellPrice": 30,
        "updatedAt": "1564232856176",
      },
    ],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
